src/ucp/rndv/rndv.c

```diff
@@ -327,7 +327,7 @@ static void ucp_rndv_complete_rma_put_zcopy(ucp_request_t *sreq, int is_frag_put
 
     if (is_frag_put) {
         ucp_send_request_id_release(sreq);
-    } else {
+    } else if (status == UCS_OK) {
```

Comment:
where will we destroy rkey?

Reply:
It will be destroyed by a receiver as a part of error flow, i.e. purging of uncompleted RTR requests on UCP EP (receiver don’t release RTR even sending is completed, it waits for ATP, or RNDV data completion)

Reply:
AFAIR you mean receiver side (whoever sends the RTR), what about the sender side (which receives the RTS and failed to put_zcopy) destroying its key?

Reply:
it will be released here anyway:
https://github.com/openucx/ucx/blob/1d356f658d2390eced7e09d417edf43b9b78fc17/src/ucp/rndv/rndv.c#L344

Reply:
but this is the sender buffer itself
what about the remote key it unpacked from RTR message?

Reply:
> what about the remote key it unpacked from RTR message?

it is released in RTR handler:
https://github.com/openucx/ucx/blob/1d356f658d2390eced7e09d417edf43b9b78fc17/src/ucp/rndv/rndv.c#L1937
after it was unpacked:
https://github.com/openucx/ucx/blob/1d356f658d2390eced7e09d417edf43b9b78fc17/src/ucp/rndv/rndv.c#L1888

Reply:
what if it's skipped here:
https://github.com/openucx/ucx/blob/1d356f658d2390eced7e09d417edf43b9b78fc17/src/ucp/rndv/rndv.c#L1932

Reply:
@brminich @yosefe nice catch, we have to release RKEY here anyway, just not send ATP to a peer. sorry for misunderstanding.
fixed.